File: solutions/16SortAnArray/java/SortAnArray.java
class SortAnArray {
  public int[] sortArray(int[] nums) {
    heapSort(nums);
    return nums;
  }

  public void heapify(int[] nums, int n, int i) {
    int largest = i;
    int left = 2 * i + 1, right = 2 * i + 2;
    if (left < n && nums[left] > nums[largest]) {
      largest = left;
    }
    if (right < n && nums[right] > nums[largest]) {
      largest = right;
    }
    if (largest != i) {
      int tmp = nums[largest];
      nums[largest] = nums[i];
      nums[i] = tmp;
      heapify(nums, n, largest);
    }
  }

  public int[] heapSort(int[] nums) {
    int n = nums.length;
    for (int i = n / 2 - 1; i > -1; i--) {
      heapify(nums, n, i);
    }
    for (int i = n - 1; i > 0; i--) {
      int tmp = nums[i];
      nums[i] = nums[0];
      nums[0] = tmp;
      heapify(nums, i, 0);
    }
    return nums;
  }

  public int[] quickSort(int[] nums, int low, int high) {
    if (low < high) {
      int pivot_i = partition(nums, low, high);
      quickSort(nums, low, pivot_i - 1);
      quickSort(nums, pivot_i + 1, high);
    }
    return nums;
  }

  public int partition(int[] nums, int low, int high) {
    int pivot = nums[low];
    int i = low, j = high;
    while (i < j) {
      while (i < j && nums[j] >= pivot) {
        j--;
      }
      while (i < j && nums[i] <= pivot) {
        i++;
      }
      int tmp = nums[i];
      nums[i] = nums[j];
      nums[j] = tmp;
    }
    int tmp = nums[j];
    nums[j] = nums[low];
    nums[low] = tmp;
    return j;
  }
}

File: solutions/16SortAnArray/python/SortAnArray.py
class Solution(object):
    def sortArray(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        return self.shellSort(nums)

    def shellSort(self, nums):
        size = len(nums)
        gap = size // 2
        while gap > 0:
            for i in range(gap, size):
                tmp = nums[i]
                j = i
                while j >= gap and nums[j - gap] > tmp:
                    nums[j] = nums[j - gap]
                    j -= gap
                nums[j] = tmp
            gap = gap // 2
        return nums

    def mergeSort(self, nums):
        if len(nums) <= 1:
            return nums

        mid = len(nums) // 2
        left_nums = self.mergeSort(nums[0:mid])
        right_nums = self.mergeSort(nums[mid:])
        return self.merge(left_nums, right_nums)

    def merge(self, left_nums, right_nums):
        nums = []
        left_i, right_i = 0, 0
        while left_i < len(left_nums) and right_i < len(right_nums):
            if left_nums[left_i] < right_nums[right_i]:
                nums.append(left_nums[left_i])
                left_i += 1
            else:
                nums.append(right_nums[right_i])
                right_i += 1

        while left_i < len(left_nums):
            nums.append(left_nums[left_i])
            left_i += 1

        while right_i < len(right_nums):
            nums.append(right_nums[right_i])
            right_i += 1

        return nums

    def quickSort(self, nums, low, high):
        if low < high:
            pivot_i = self.partition(nums, low, high)
            self.quickSort(nums, low, pivot_i - 1)
            self.quickSort(nums, pivot_i + 1, high)
        return nums

    def partition(self, nums, low, high):
        pivot = nums[low]
        i, j = low, high
        while i < j:
            while i < j and nums[j] >= pivot:
                j -= 1
            while i < j and nums[i] <= pivot:
                i += 1
            nums[i], nums[j] = nums[j], nums[i]

        nums[j], nums[low] = nums[low], nums[j]
        return j
